README
LNMP一键安装包 - Readme

LNMP一键安装包是什么?

LNMP一键安装包是一个用Linux Shell编写的可以为CentOS/RHEL/Fedora/Aliyun/Amazon、Debian/Ubuntu/Raspbian/Deepin/Mint Linux VPS或独立主机安装LNMP(Nginx/MySQL/PHP)、LNMPA(Nginx/MySQL/PHP/Apache)、LAMP(Apache/MySQL/PHP)生产环境的Shell程序。

LNMP一键安装包有哪些功能？

支持自定义Nginx、PHP编译参数及网站和数据库目录、支持生成LetseEcrypt证书、LNMP模式支持多PHP版本、支持单独安装Nginx/MySQL/MariaDB/Pureftpd服务器，同时提供一些实用的辅助工具如：虚拟主机管理、FTP用户管理、Nginx、MySQL/MariaDB、PHP的升级、常用缓存组件Redis/Xcache等的安装、重置MySQL root密码、502自动重启、日志切割、SSH防护DenyHosts/Fail2Ban、备份等许多实用脚本。

LNMP官网：https://lnmp.org
作者: licess admin@lnmp.org
问题反馈&技术支持论坛：https://bbs.vpser.net/forum-25-1.html
打赏捐赠：https://lnmp.org/donation.html

LNMP安装

安装前确认已经安装wget命令，如提示wget: command not found ，使用 yum install wget  或  apt-get install wget  命令安装。
为防止掉线等情况，建议使用screen，可以先执行：screen -S lnmp 命令后，再执行LNMP安装命令：

wget http://soft.vpser.net/lnmp/lnmp1.8.tar.gz -cO lnmp1.8.tar.gz && tar zxf lnmp1.8.tar.gz && cd lnmp1.8 && ./install.sh {lnmp|lnmpa|lamp} 

如断线可使用 screen -r lnmp  恢复。详细安装教程参考：https://lnmp.org/install.html

常用功能说明

以下操作需在lnmp安装包目录下执行，如lnmp1.8

自定义参数
lnmp.conf配置文件，可以修改lnmp.conf自定义下载服务器地址、网站/数据库目录及添加nginx模块和php编译参数；不论安装升级都会调用该文件里的设置(如果修改了默认的参数建议备份此文件)；
详细参数说明可以 https://lnmp.org/faq/lnmp-software-list.html#lnmp.conf 查看；

FTP服务器
执行： ./pureftpd.sh  安装，可使用  lnmp ftp {add|list|del}  进行管理。

升级脚本：
执行： ./upgrade.sh  按提示进行选择
也可以直接带参数：./upgrade.sh {nginx|mysql|mariadb|php|phpa|mphp|m2m|phpmyadmin}
* 参数: nginx   可升级至任意Nginx版本。
* 参数: mysql   可升级至任意MySQL版本，MySQL升级风险较大，虽然会自动备份数据，依然建议自行再备份一下。
* 参数: mariadb 可升级已安装的Mariadb，虽然会自动备份数据，依然建议自行再备份一下。
* 参数: m2m     可从MySQL升级至Mariadb，虽然会自动备份数据，依然建议自行再备份一下。
* 参数: php     仅适用于LNMP，可升级至大部分PHP版本。
* 参数: phpa    可升级LNMPA/LAMP的PHP至大部分版本。
* 参数: mphp    多PHP版本升级工具，只支持7.2.x-7.2.x类似小版本升级，大版本直接新装即可；
* 参数: phpmyadmin    可升级phpMyadmin。

扩展插件
执行:  ./addons.sh {install|uninstall} {eaccelerator|xcache|memcached|opcache|redis|apcu|imagemagick|ioncube} 
以下为扩展插件安装使用说明
缓存加速：
* 参数: xcache 安装时需选择版本和设置密码，http://yourIP/xcache/ 进行管理，用户名 admin，密码为安装xcache时设置的。
* 参数: redis  安装redis
* 参数: memcached 可选择php-memcache或php-memcached扩展。
* 参数: opcache 可访问 http://yourIP/ocp.php 进行管理。
* 参数: eaccelerator 安装。
* 参数: apcu 安装apcu php扩展，支持php7，可访问 http://yourIP/apc.php 进行管理。 
**请勿安装多个缓存类扩展模块，多个可能导致网站出现问题 ！**

图像处理：
imageMagick安装卸载执行： ./addons.sh {install|uninstall} imageMagick imageMagick路径：/usr/local/imagemagick/bin/ 。

解密：
IonCube安装执行： ./addons.sh {install|uninstall} ionCube 。

其他常用脚本：
* 可选1，多PHP版本安装执行： ./install.sh mphp  可以安装多个PHP版本 ，只支持LNMP模式，lnmp vhost add时进行选择或使用时需要将nginx虚拟主机配置文件里的include enable-php.conf替换为 include enable-php5.6.conf 即可前面的5.6换成你刚才安装的PHP的大版本号5.* 或7.0之类的。
* 可选2，数据库安装执行： ./install.sh db  可以直接单独安装MySQL或MariaDB数据库。
* 可选3，Nginx安装执行： ./install.sh nginx 可以直接单独安装Nginx。
以下工具在lnmp安装包tools目录下，可拷贝到其他目录下运行
* 可选4，执行： ./reset_mysql_root_password.sh  可重置MySQL/MariaDB的root密码。
* 可选5，执行： ./check502.sh   可检测php-fpm是否挂掉,502报错时重启，配合crontab使用。
* 可选6，执行： ./cut_nginx_logs.sh  日志切割脚本。
* 可选7，执行： ./remove_disable_function.sh  运行此脚本可删掉禁用函数。

无人值守安装
请访问 https://lnmp.org/faq/v1-5-auto-install.html 查看无人值守安装教程

卸载

卸载LNMP、LNMPA或LAMP可执行： ./uninstall.sh  按提示选择即可卸载。

状态管理
* LNMP/LNMPA/LMAP状态管理： lnmp {start|stop|reload|restart|kill|status} 
* Nginx状态管理： lnmp nginx或/etc/init.d/nginx {start|stop|reload|restart} 
* MySQL状态管理： lnmp mysql或/etc/init.d/mysql {start|stop|restart|reload|force-reload|status} 
* MariaDB状态管理： lnmp mariadb或/etc/init.d/mariadb {start|stop|restart|reload|force-reload|status} 
* PHP-FPM状态管理： lnmp php-fpm或/etc/init.d/php-fpm {start|stop|quit|restart|reload|logrotate} 
* PureFTPd状态管理： lnmp pureftpd或/etc/init.d/pureftpd {start|stop|restart|kill|status} 
* Apache状态管理： lnmp httpd或/etc/init.d/httpd {start|stop|restart|graceful|graceful-stop|configtest|status} 

虚拟主机管理
* 添加： lnmp vhost add 
* 删除： lnmp vhost del 
* 列出： lnmp vhost list 
* 数据库管理： lnmp database {add|list|edit|del} 
* FTP用户管理： lnmp ftp {add|list|edit|del|show} 
* SSL添加： lnmp ssl add
* 通配符/泛域名SSL添加：lnmp dnsssl {cx|ali|cf|dp|he|gd|aws} 需依赖域名dns api

相关图形界面
* PHPMyAdmin：http://yourIP/phpmyadmin/
* phpinfo：http://yourIP/phpinfo.php
* PHP探针：http://yourIP/p.php
* Xcache管理界面：http://yourIP/xcache/
* Zend Opcache管理界面：http://yourIP/ocp.php
* apcu管理界面：http://yourIP/apc.php

LNMP相关目录文件

#目录位置
* Nginx：/usr/local/nginx/
* MySQL：/usr/local/mysql/
* MariaDB：/usr/local/mariadb/
* PHP：/usr/local/php/
* 多PHP目录：/usr/local/php5.6/ 版本号随安装版本不同而不同
* PHP扩展插件配置文件目录：/usr/local/php/conf.d/
* PHPMyAdmin：/www/web/default/phpmyadmin/
* 默认虚拟主机网站目录：/www/web/default/
* Nginx日志目录：/www/log/

#配置文件：
* Nginx主配置文件：/usr/local/nginx/conf/nginx.conf
* MySQL/MariaDB配置文件：/etc/my.cnf
* PHP配置文件：/usr/local/php/etc/php.ini
* PHP-FPM配置文件：/usr/local/php/etc/php-fpm.conf
* PureFtpd配置文件：/usr/local/pureftpd/etc/pure-ftpd.conf
* Apache配置文件：/usr/local/apache/conf/httpd.conf

技术支持

技术支持论坛：https://bbs.vpser.net/forum-25-1.html

ChangeLog
更新记录

LNMP 1.8

增加PHP 8.0支持；
增加Oracle Linux支持；
增加WSL优化支持；
增加CentOS6源自动调整；
增加BuyPass、ZeroSSL SSL证书；
增加php-memcache支持PHP 8.0；
增加imagick支持PHP 8.0；
增加apcu支持PHP 8.0；
优化时间同步；
优化Aliyunx Linux优化；
优化sudo下添加虚拟主机SSL；
优化CentOS8 Stream支持；
优化PHP下载；
更新诸多软件版本；
其他一些功能优化及调整......

LNMP 1.7

增加PHP 7.4支持；
增加MariaDB 10.4支持；
增加systemd service脚本；
增加phpmyadmin 5.0；
增加多PHP版本升级工具 ./upgrade.sh mphp，只支持7.2.x-7.2.x类似小版本升级，大版本直接新装即可；
移除MariaDB 10.0；
优化lnmp管理脚本；
优化数据库root密码重置工具；
优化系统版本获取脚本；
优化redis安装脚本；
优化fail2ban安装脚本；
优化nginx安装、升级脚本；
完善CentOS8、Ubuntu20、Deepin20、Fedora等新版本的支持；
完善MySQL升级时增加旧版本下载镜像；
调整php7 memcached使用稳定版；
修复lamp模式下ssl chain添加的bug；
更新诸多软件版本；
其他一些功能优化及调整......

LNMP 1.6

增加PHP 7.3支持；
增加MariaDB 10.3支持；
增加Nginx/Apache TLS 1.3支持；
增加使用DNS API方式只创建SSL证书不添加网站，命令lnmp onlyssl；
增加Fedora 30、Ubuntu 19.04、RHEL 8等新版本的支持；
增加nginx、apache等一些例子，并移至conf/example目录下；
增加在低于2048MB内存是增加2GB SWAP或低于1024MB时增加1GB SWAP的选项；
增加反向代理配置例子文件 nginx-reverse-proxy-example.conf；
增加只安装nginx模式增加拷贝lnmp管理工具；
增加 RHELRepo=local 参数，RHEL安装时使用RHEL本地源，不设置源为163 centos源；
增加 CheckMirror=n 参考，安装时不检查下载镜像，方便无网络安装；
优化不安装数据库时的处理流程；
优化acme.sh规则可能不生效的情况；
优化nginx系统满足条件是启用reuseport；
优化程序代码目录清理；
优化重写nginx管理脚本；
优化安装数据库时的内存判断，放宽内存限制；
修复新版深度Deepin 15.8桌面版问题；
修复新版Fedora 29下问题；
调整mysql使用带boost源码，同时支持不带boost安装方式；
调整卸载时不删除安装时添加的iptables规则；
调整fileinfo安装选项，内存满足时自动安装(可关闭)；
升级Nginx至1.16.0；
升级各程序版本；
更新诸多软件版本；
其他一些功能优化及调整......

LNMP 1.5

增加PHP 7.2支持；
增加MariaDB 10.2支持；
增加MySQL 8.0支持；
增加PHP模块fileinfo选项，安装或升级时可以选择是否安装开启，选项在安装包目录下 lnmp.conf 中 Enable_PHP_Fileinfo 默认为 n 不开启，如有安装设置为 y ；
增加Nginx lua模块选项，安装或升级时可以选择是否为nginx安装开启lua，选项在安装包目录下 lnmp.conf 中 Enable_Nginx_Lua 默认为 n 不开启，如有安装设置为 y ；
增加thinkphp、yii2、codeigniter、laravel伪静态；
增加使用acme.sh生成SSL证书，移除官方的certbot；
增加Nginx编译参数，默认开启stream；
增加Apache 2.4支持http2；
增加旧版本Ubuntu中国地区源；
增加虚拟主机添加时pathinfo可选项；
增加certbot升级到acme.sh的一键升级功能命令为 ./upgrade1.x-1.5.sh ssl 按提示输入和原来SSL虚拟主机一样配置的信息就可以；
增加多核编译加快安装时间；
增加kali Linux发行版支持；
增加对Amazone Linux 2支持；
增加修复MySQL gcc7 bug；
增加magento2实例配置文件 magento2-example.conf；
增加无人值守安装，官网上有具体教程；
增加LetseEcrypt 通配符/泛域名 SSL证书生成，官网上有具体教程；
优化MySQL/MariaDB root密码重装工具；
修复深度Deepin桌面版问题；
修复MySQL bug；
修复MariaDB bug；
移除PHP禁用函数scandir；
调整默认PHP版本为5.6；
调整默认Apache版本为2.4；
调整MySQL/MariaDB安装时未输入密码由原来密码设为 root 调整为 lnmp.org#随机数字 ；
调整MySQL国内下载源为科大的；
调整LAMP模式下SSL添加过程增加中间证书可选项；
调整默认虚拟主机页面字体显示等细节；
调整nginx伪静态至/usr/local/nginx/conf/rewrite目录；
调整优化lnmp下载源选择；
完善卸载脚本；
完善优化lnmp升级脚本 upgrade1.x-1.5.sh；
调整数据库内存优化加载方法；
其他一些功能优化及调整......

LNMP 1.4

增加LNMP模式下多PHP版本安装，安装包目录下运行：./install.sh mphp 进行安装；
增加单独nginx安装，安装包目录下运行：./install.sh nginx 进行安装；
增加单独数据库安装，安装包目录下运行：./install.sh db 进行安装；
增加Let's Encrypt自动生成证书和自备SSL证书支持；
增加不安装数据库选项；
增加多PHP版本下扩展安装支持；
增加ioncube php7的支持；
增加apcu php扩展；
增加PHP7.1支持；
增加lnmp管理脚本升级脚本，安装包目录下运行：./upgrade1.x-1.4.sh；
增加对中国用户的识别，使用国内源；
增加Amazon Linux、Linux Mint支持；
增加计算显示LNMP安装使用时间；
增加是否已安装LNMP判断，防止重复安装覆盖数据；
增加在已存在目录下安装数据库时先进行备份；
增加防跨目录配置文件移除脚本，安装包目录里tools目录下运行 ./remove_open_basedir_restriction.sh ；
lnmp管理脚本增加SSL选项，可以选择Let's Encrypt或自定义证书和key文件，Letsencrypt每7天自动尝试续期；
lnmp管理脚本增加ftp账号信息显示；
lnmp管理脚本增加LNMP模式下添加虚拟主机时多PHP版本选择选项；
lnmp管理脚本去除add more domain name提示，添加直接输入域名，不添加直接回车；
lnmp管理脚本添加虚拟主机处理流程上的优化；
lnmp管理脚本增加多PHP状态管理支持；
lnmp管理脚本增加 lnmp ftp show 显示ftp账号信息；
lnmp管理脚本已存在虚拟主机SSL时可以直接lnmp ssl add 进行添加；
nginx默认使用openssl 1.0.2进行安装，以便更好的支持http2；
修复完善国内composer安装问题，国内能正常流畅的安装；
修复lnmpa/lamp模式下https问题；
修复arm下centos不存在libcurl依赖包的问题；
调整部分软件源码包格式为tar.bz2；
调整apache日志LogLevel为crit；
调整php模块的引入方式,使用with-config-file-scan-dir引入；
调整默认PHP版本调整为PHP 5.5；
调整国内外节点顺序，1,3,5 为国内节点(soft1.vpser.net 可调整数字)，2,4,6,8 为国外节点(soft2.vpser.net 可调整数字)。
调整默认虚拟主机不在使用www.lnmp.org域名，使用_代替；
增加lnmp.conf配置文件中nginx是否使用新版openssl编译安装选项；
增加LNMP模式下上传目录或不需php解析目录php禁止访问的实例代码，去掉注释修改其中目录重启nginx即可使用；
增加LNMP模式下的https实例配置文件：/usr/local/nginx/conf/enable-ssl-example.conf；
增加LNMPA模式下的https实例配置文件：/usr/local/nginx/conf/enable-lnmpa-ssl-vhost-example.conf；
增加LAMP模式下的https实例配置文件：/usr/local/apache/conf/enable-apache-ssl-vhost-example.conf；
增加PHP模块zts和non-zts自动识别；
添加Ubuntu 17.04支持；
php探针在php7下优化；
fastcgi.conf增加防跨目录的二次设置，移除参考：https://lnmp.org/faq/lnmp-vhost-add-howto.html#user.ini ；
imagemagick增加webp格式支持；
去掉pureftp配置文件匿名模式下匿名用户上传权限；
防止误卸载，卸载脚本增加备份数据库目录；
优化pureftpd安装脚本，可单独在服务器进行安装；
优化nginx虚拟主机配置文件，增加对let'sencrypt目录的允许规则；
优化卸载脚本；
优化组件模块安装脚本；
优化部分启动脚本；
调整升级MySQL/MariaDB时my.cnf和启动脚本的位置到MySQL/MariaDB备份目录下；
调整优化fail2ban、denyhosts安装脚本；
诸多小的优化及调整......

LNMP 1.3

优化安装脚本、升级脚本及组件安装脚本；
增加lnmp.conf配置文件，可以自定义下载服务器地址、数据库目录及nginx和php编译参数；不论安装升级都会调用该文件里的设置；
增加安装和升级时会根据配置自动优化相关参数；
增加php7支持；
增加mysql 5.7支持；
增加phpmyadmin升级功能；
php增加Composer工具；
php 5.4以上版本增加intl、xsl扩展；
增加nginx升级到1.9.5以上版本支持http2；
增加php-redis、php-memcached、php-memcache、imagick PHP7下模块的支持；
增加升级PHP/MySQL/MariaDB时对源目录进行完整备份，防范升级失败风险；
增加卸载LNMP/LNMPA/LAMP时对数据库目录进行备份功能；
增加lnmp脚本创建虚拟主机时可选创建MySQL数据库和FTP账号选项(需先安装好pureftpd)；
增加安装或升级完清理目录节约硬盘空间；
redis仅保留最新版，删除旧版支持；
修正在部分系统下MySQL无法正确设置密码的问题；
修正完善防火墙规则；
pureftpd去掉MySQL的依赖和web面板，只使用lnmp ftp {add|list|edit|del} 进行ftp账号管理；
调整pureftpd配置文件目录至/usr/local/pureftpd/etc/下；
nginx.conf php启用配置简化为include enable-php.conf;
诸多小的更新及调整......


2015年6月1日发布LNMP 1.2

完全重写优化安装脚本、升级脚本及组件安装等大部分脚本；
状态管理、虚拟主机管理、FTP管理及数据库管理全部采用lnmp命令替换，去掉/root/vhost.sh和/root/lnmp脚本；
Nginx、MySQL/MariaDB、PHP的升级脚本整合为upgrade.sh;
缓存、加速、imageMagick、ionCube等的安装卸载采用addons.sh进行；
增加SSH防护工具DenyHosts、Failban安装脚本及DenyHosts被封IP删除脚本；
增加新的备份脚本backup.sh；
支持的Linux发行版：CentOS 5-7、RHEL 6-7、Fedora 20-21、Aliyun、Debian 6-8、Raspbian、Ubuntu 10.10-15.04；
支持ARM平台，Raspberry Pi已测试；
MySQL安装时可选5.1.73、5.5.42、5.6.23，MariaDB可选5.5.42、10.0.17；
PHP安装时可选5.2.17、5.3.29、5.4.41、5.5.25、5.6.9；
安装时增加内存优化jemalloc或gperftools；
Apache安装时可选2.2.29或2.4.12；
Apache日志统一到/www/log/下；
更新Nginx至1.8.0；
更新Zend Guard Loader至支持PHP 5.6；
Nginx配置文件进行了调整，启用php和pathinfo采用include方式，增加ssl范例配置；



2014年6月1日发布LNMP 1.1

调整优化脚本安装脚本；
安装时增加MariaDB选项；
增加MySQL升级脚本 upgrade_mysql.sh；
增加MySQL升级到MariaDB脚本 upgrade_mysql2mariadb.sh；
增加MariaDB升级脚本 upgrade_mariadb.sh；
增加Redis安装脚本 redis.sh；
增加OPcache安装脚本 opcache.sh；
增加MySQL/MariaDB root密码重置脚本；
添加pathinfo支持，去掉include pathinfo的#，在try_files前加#重启nginx即可使用；
重写nginx init.d脚本；
优化PHP升级脚本；

addons.sh
#!/usr/bin/env bash
export PATH=$PATH:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:~/bin

# Check if user is root
if [ $(id -u) != "0" ]; then
    echo "Error: You must be root to run this script"
    exit 1
fi

cur_dir=$(pwd)
action=$1
action2=$2

. lnmp.conf
. include/main.sh
. include/init.sh
. include/version.sh
. include/eaccelerator.sh
. include/xcache.sh
. include/memcached.sh
. include/opcache.sh
. include/redis.sh
. include/imageMagick.sh
. include/ionCube.sh
. include/apcu.sh

Display_Addons_Menu()
{
    echo "##### cache / optimizer / accelerator #####"
    echo "1: eAccelerator"
    echo "2: XCache"
    echo "3: Memcached"
    echo "4: opcache"
    echo "5: Redis"
    echo "6: apcu"
    echo "##### Image Processing #####"
    echo "7: imageMagick"
    echo "##### encryption/decryption utility for PHP #####"
    echo "8: ionCube Loader"
    echo "exit: Exit current script"
    echo "#####################################################"
    read -p "Enter your choice (1, 2, 3, 4, 5, 6, 7, 8 or exit): " action2
}

Restart_PHP()
{
    if [ -s /usr/local/apache/bin/httpd ] && [ -s /usr/local/apache/conf/httpd.conf ] && [ -s /etc/init.d/httpd ]; then
        echo "Restarting Apache......"
        /etc/init.d/httpd restart
    else
        echo "Restarting php-fpm......"
        ${PHPFPM_Initd} restart
    fi
}

clear
echo "+-----------------------------------------------------------------------+"
echo "|            Addons script for LNMP V1.8, Written by Licess             |"
echo "+-----------------------------------------------------------------------+"
echo "|    A tool to Install cache,optimizer,accelerator...addons for LNMP    |"
echo "+-----------------------------------------------------------------------+"
echo "|           For more information please visit https://lnmp.org          |"
echo "+-----------------------------------------------------------------------+"

Select_PHP()
{
    if [[ ! -s /usr/local/php5.2/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php5.2.conf ]] && [[ ! -s /usr/local/php5.3/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php5.3.conf ]] && [[ ! -s /usr/local/php5.4/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php5.4.conf ]] && [[ ! -s /usr/local/php5.5/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php5.5.conf ]] && [[ ! -s /usr/local/php5.6/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php5.6.conf ]] && [[ ! -s /usr/local/php7.0/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php7.0.conf ]] && [[ ! -s /usr/local/php7.1/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php7.1.conf ]] && [[ ! -s /usr/local/php7.2/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php7.2.conf ]] && [[ ! -s /usr/local/php7.3/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php7.3.conf ]] && [[ ! -s /usr/local/php7.4/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php7.4.conf ]] && [[ ! -s /usr/local/php8.0/sbin/php-fpm && ! -s /usr/local/nginx/conf/enable-php8.0.conf ]]; then
        PHP_Path='/usr/local/php'
        PHPFPM_Initd='/etc/init.d/php-fpm'
    else
        echo "Multiple PHP version found, Please select the PHP version."
        Cur_PHP_Version="`/usr/local/php/bin/php-config --version`"
        Echo_Green "1: Default Main PHP ${Cur_PHP_Version}"
        if [[ -s /usr/local/php5.2/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php5.2.conf && -s /etc/init.d/php-fpm5.2 ]]; then
            Echo_Green "2: PHP 5.2 [found]"
        fi
        if [[ -s /usr/local/php5.3/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php5.3.conf && -s /etc/init.d/php-fpm5.3 ]]; then
            Echo_Green "3: PHP 5.3 [found]"
        fi
        if [[ -s /usr/local/php5.4/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php5.4.conf && -s /etc/init.d/php-fpm5.4 ]]; then
            Echo_Green "4: PHP 5.4 [found]"
        fi
        if [[ -s /usr/local/php5.5/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php5.5.conf && -s /etc/init.d/php-fpm5.5 ]]; then
            Echo_Green "5: PHP 5.5 [found]"
        fi
        if [[ -s /usr/local/php5.6/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php5.6.conf && -s /etc/init.d/php-fpm5.6 ]]; then
            Echo_Green "6: PHP 5.6 [found]"
        fi
        if [[ -s /usr/local/php7.0/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php7.0.conf && -s /etc/init.d/php-fpm7.0 ]]; then
            Echo_Green "7: PHP 7.0 [found]"
        fi
        if [[ -s /usr/local/php7.1/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php7.1.conf && -s /etc/init.d/php-fpm7.1 ]]; then
            Echo_Green "8: PHP 7.1 [found]"
        fi
        if [[ -s /usr/local/php7.2/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php7.2.conf && -s /etc/init.d/php-fpm7.2 ]]; then
            Echo_Green "9: PHP 7.2 [found]"
        fi
        if [[ -s /usr/local/php7.3/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php7.3.conf && -s /etc/init.d/php-fpm7.3 ]]; then
            Echo_Green "10: PHP 7.3 [found]"
        fi
        if [[ -s /usr/local/php7.4/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php7.4.conf && -s /etc/init.d/php-fpm7.4 ]]; then
            Echo_Green "11: PHP 7.4 [found]"
        fi
        if [[ -s /usr/local/php8.0/sbin/php-fpm && -s /usr/local/nginx/conf/enable-php8.0.conf && -s /etc/init.d/php-fpm8.0 ]]; then
            Echo_Green "12: PHP 8.0 [found]"
        fi
        Echo_Yellow "Enter your choice (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 or 12 ): "
        read php_select
        case "${php_select}" in
            1)
                echo "Current selection: PHP ${Cur_PHP_Version}"
                PHP_Path='/usr/local/php'
                PHPFPM_Initd='/etc/init.d/php-fpm'
                ;;
            2)
                echo "Current selection: PHP `/usr/local/php5.2/bin/php-config --version`"
                PHP_Path='/usr/local/php5.2'
                PHPFPM_Initd='/etc/init.d/php-fpm5.2'
                ;;
            3)
                echo "Current selection: PHP `/usr/local/php5.3/bin/php-config --version`"
                PHP_Path='/usr/local/php5.3'
                PHPFPM_Initd='/etc/init.d/php-fpm5.3'
                ;;
            4)
                echo "Current selection: PHP `/usr/local/php5.4/bin/php-config --version`"
                PHP_Path='/usr/local/php5.4'
                PHPFPM_Initd='/etc/init.d/php-fpm5.4'
                ;;
            5)
                echo "Current selection: PHP `/usr/local/php5.5/bin/php-config --version`"
                PHP_Path='/usr/local/php5.5'
                PHPFPM_Initd='/etc/init.d/php-fpm5.5'
                ;;
            6)
                echo "Current selection: PHP `/usr/local/php5.6/bin/php-config --version`"
                PHP_Path='/usr/local/php5.6'
                PHPFPM_Initd='/etc/init.d/php-fpm5.6'
                ;;
            7)
                echo "Current selection:: PHP `/usr/local/php7.0/bin/php-config --version`"
                PHP_Path='/usr/local/php7.0'
                PHPFPM_Initd='/etc/init.d/php-fpm7.0'
                ;;
            8)
                echo "Current selection:: PHP `/usr/local/php7.1/bin/php-config --version`"
                PHP_Path='/usr/local/php7.1'
                PHPFPM_Initd='/etc/init.d/php-fpm7.1'
                ;;
            9)
                echo "Current selection:: PHP `/usr/local/php7.2/bin/php-config --version`"
                PHP_Path='/usr/local/php7.2'
                PHPFPM_Initd='/etc/init.d/php-fpm7.2'
                ;;
            10)
                echo "Current selection:: PHP `/usr/local/php7.3/bin/php-config --version`"
                PHP_Path='/usr/local/php7.3'
                PHPFPM_Initd='/etc/init.d/php-fpm7.3'
                ;;
            11)
                echo "Current selection:: PHP `/usr/local/php7.4/bin/php-config --version`"
                PHP_Path='/usr/local/php7.4'
                PHPFPM_Initd='/etc/init.d/php-fpm7.4'
                ;;
            12)
                echo "Current selection:: PHP `/usr/local/php8.0/bin/php-config --version`"
                PHP_Path='/usr/local/php8.0'
                PHPFPM_Initd='/etc/init.d/php-fpm8.0'
                ;;
            *)
                echo "Default,Current selection: PHP ${Cur_PHP_Version}"
                php_select="1"
                PHP_Path='/usr/local/php'
                PHPFPM_Initd='/etc/init.d/php-fpm'
                ;;
        esac
    fi
}

Addons_Get_PHP_Ext_Dir()
{
    Cur_PHP_Version="`${PHP_Path}/bin/php-config --version`"
    zend_ext_dir="`${PHP_Path}/bin/php-config --extension-dir`/"
}

if [[ "${action}" == "" || "${action2}" == "" ]]; then
    action='install'
    Display_Addons_Menu
fi
Get_Dist_Name
Select_PHP

    case "${action}" in
    install)
        case "${action2}" in
            1|e[aA]ccelerator)
                Install_eAccelerator
                ;;
            2|[xX]cache)
                Install_XCache
                ;;
            3|[mM]emcached)
                Install_Memcached
                ;;
            4|opcache)
                Install_Opcache
                ;;
            5|[rR]edis)
                Install_Redis
                ;;
            6|apcu)
                Install_Apcu
                ;;
            7|image[mM]agick)
                Install_ImageMagic
                ;;
            8|ion[cC]ube)
                Install_ionCube
                ;;
            [eE][xX][iI][tT])
                exit 1
                ;;
            *)
                echo "Usage: ./addons.sh {install|update|uninstall} {eaccelerator|xcache|memcached|opcache|redis|imagemagick|ioncube}"
                ;;
        esac
        ;;
    update)
        case "${action2}" in
            e[aA]ccelerator)
                Uninstall_eAccelerator
                Install_eAccelerator
                ;;
            [xX]cache)
                Uninstall_XCache
                Install_XCache
                ;;
            [mM]emcached)
                Uninstall_Memcached
                Install_Memcached
                ;;
            opcache)
                Uninstall_Opcache
                Install_Opcache
                ;;
            [rR]edis)
                Uninstall_Redis
                Install_Redis
                ;;
            apcu)
                Uninstall_Apcu
                Install_Apcu
                ;;
            image[mM]agick)
                Uninstall_ImageMagick
                Install_ImageMagic
                ;;
            ion[cC]ube)
                Uninstall_ionCube
                Install_ionCube
                ;;
            *)
                echo "Usage: ./addons.sh {install|update|uninstall} {eaccelerator|xcache|memcached|opcache|redis|apcu|imagemagick|ioncube|nghttp2}"
                ;;
        esac
        ;;
    uninstall)
        case "${action2}" in
            e[aA]ccelerator)
                Uninstall_eAccelerator
                ;;
            [xX]cache)
                Uninstall_XCache
                ;;
            [mM]emcached)
                Uninstall_Memcached
                ;;
            opcache)
                Uninstall_Opcache
                ;;
            [rR]edis)
                Uninstall_Redis
                ;;
            apcu)
                Uninstall_Apcu
                ;;
            image[mM]agick)
                Uninstall_ImageMagick
                ;;
            ion[cC]ube)
                Uninstall_ionCube
                ;;
            *)
                echo "Usage: ./addons.sh {install|update|uninstall} {eaccelerator|xcache|memcached|opcache|redis|apcu|imagemagick|ioncube}"
                ;;
        esac
        ;;
    [eE][xX][iI][tT])
        exit 1
        ;;
    *)
        echo "Usage: ./addons.sh {install|update|uninstall} {eaccelerator|xcache|memcached|opcache|redis|apcu|imagemagick|ioncube}"
        exit 1
        ;;
    esac
